### PR TITLE
streamingccl: fix span use-after-finish in ingestion frontier

### LINKED ISSUE #84054: ccl/streamingccl/streamingest: TestTenantStreaming failed
ccl/streamingccl/streamingest.TestTenantStreaming [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5685691?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5685691?buildTab=artifacts#/) on master @ [88789feea153b61b379c0240a07482bdfaf0c932](https://github.com/cockroachdb/cockroach/commits/88789feea153b61b379c0240a07482bdfaf0c932):

```
=== RUN   TestTenantStreaming
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreaming2237887931
    test_log_scope.go:80: use -show-logs to present logs inline
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/tenant-streaming
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreaming.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-17431

### PR DESCRIPTION
**Pull request by stevendanna**
This fixes the following use-after-finish panic:

    panic: use of Span after Finish. Span: ingestfntr. Finish previously
    called at: <stack not captured. Set debugUseAfterFinish>

    goroutine 1617744 [running]:
    github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc002c5f180)
    	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:186 +0x279
    github.com/cockroachdb/cockroach/pkg/util/tracing.(*Tracer).startSpanGeneric(0xc0138b2a50, {0x72291c0, 0xc019ebbe40}, {0x68e4eb2, 0x1d}, {{0x0}, 0x0, {0x0, 0x0, {{0x0, ...}, ...}, ...}, ...})
    	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1207 +0x997
    github.com/cockroachdb/cockroach/pkg/util/tracing.(*Tracer).StartSpanCtx(0xc0138b2a50, {0x72291c0, 0xc019ebbe40}, {0x68e4eb2, 0x1d}, {0xc010cbb760, 0x1, 0x1})
    	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1062 +0x1a7
    github.com/cockroachdb/cockroach/pkg/util/tracing.ChildSpan({0x72291c0, 0xc019ebbe40}, {0x68e4eb2, 0x1d})
    	github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1577 +0x145
    github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient.(*partitionedStreamClient).Heartbeat(0xc00b311080, {0x72291c0, 0xc019ebbe40}, 0xac93aa873318001, {0x16ffc388c7d7c42a, 0x0, 0x0})
    	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:83 +0xb7
    github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*heartbeatSender).maybeHeartbeat(0xc01980f880, {0x72291c0, 0xc019ebbe40}, {0x16ffc388c7d7c42a, 0x0, 0x0})
    	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:180 +0x250
    github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*heartbeatSender).startHeartbeatLoop.func1.1()
    	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:207 +0x41b
    github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*heartbeatSender).startHeartbeatLoop.func1({0x72291c0, 0xc019ebbe40})
    	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:227 +0xb5
    github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
    	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:169 +0x52
    golang.org/x/sync/errgroup.(*Group).Go.func1()
    	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:74 +0xb4
    created by golang.org/x/sync/errgroup.(*Group).Go
    	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:71 +0xdd
    I220708 05:29:42.017167 1 (gostd) testmain.go:90  [-] 1  Test //pkg/ccl/streamingccl/streamingest:streamingest_test exited with error code 2

The use after finish was caused by goroutines in the ingestion
frontier processor that lived past a call to
(*ProcessorBase).InternalClose, which finishes the span attached to
the context passed to the processor in Start.

To address this we:

- ensure that we stop our heartbeat thread before calling
  InternalClose in ConsumerClosed, and

- provide a TrailingMetaCallback so that we can perform cleaned when
  DrainHelper() is called. When a TrailingMetaCallback is provided,
  DrainHelper() calls it instead of InternalClose(), allowing us to
  correctly clean up before the span is closed.

- use context cancellation rather than a channel to control the
  heartbeat loop exit to avoid having to deal with avoiding
  double-closes of the channel.

Fixes #84054

Release note: None

### CODE REVIEW

**Comment by stevendanna** [comment]
I've run this under stress for a long time. But, take that with a grain of salt because I ran this under stress for a long time before merging the unskip of TestTenantStreaming and didn't see this until after merging.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go [hunk 1]
```diff
@@ -125,6 +125,10 @@ func newStreamIngestionFrontierProcessor(
 		nil, /* memMonitor */
 		execinfra.ProcStateOpts{
 			InputsToDrain: []execinfra.RowSource{sf.input},
+			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
```
Comment (samiskin) [inline pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:128]: Just for my own knowledge: what does `TrailingMeta` mean? When is `ConsumerClosed` ran vs this?

**Review by samiskin — APPROVED** [review]
Lgtm ✨

**Comment by stevendanna** [comment]
TFTR!

bors r=samiskin

**Pull request merged**